firehose.go

```diff
@@ -1,86 +1,191 @@
 		return true, nil
 	}
 	return false, nil
 }
 
-// Each firehose stream can support up to 2,000 transactions per second for writes,
-// 5,000 records a second, up to a maximum total data write rate of 5 MB per second by default.
-// TODO: payload inspection & throttling
-// PutRecordBatch can take up to 500 records per call or 4 MB per call, whichever is smaller.
-// http://docs.aws.amazon.com/firehose/latest/dev/limits.html
-func (p *Producer) firehoseFlush(counter *int, timer *time.Time) bool {
-	// If a second has passed since the last timer start, reset the timer
-	if time.Now().After(timer.Add(1 * time.Second)) {
-		*timer = time.Now()
-		*counter = 0
-	}
+func (p *Firehose) setConcurrency(concurrency int) {
+	p.concurrencyMu.Lock()
+	p.concurrency = concurrency
+	p.concurrencyMu.Unlock()
+}
 
-	*counter++
+func (p *Firehose) getConcurrency() (concurrency int) {
+	p.concurrencyMu.Lock()
+	concurrency = p.concurrency
+	p.concurrencyMu.Unlock()
+	return
+}
 
-	// If we have attempted 1000 times and it has been less than one second
-	// since we started sending then we need to wait for the second to finish
-	if *counter >= firehoseWritesPerSec && !(time.Now().After(timer.Add(1 * time.Second))) {
-		// Wait for the remainder of the second - timer and counter
-		// will be reset on next pass
-		<-time.After(1 * time.Second)
-	}
+func (p *Firehose) msgBufSize() int {
+	return p.getConcurrency() * 1000
+}
 
-	return true
+// Send a message to Firehose asyncronously
+func (p *Firehose) Send(msg *Message) {
+	msg.SetValue(append(msg.Value(), truncatedRecordTerminator...))
+	p.wg.Add(1)
+	go func() {
+		p.messages <- msg
+		p.wg.Done()
+	}()
+}
+
+// TryToSend tries to send the message, but if the channel is full it drops the message, and returns an error.
+func (p *Firehose) TryToSend(msg *Message) error {
+	msg.SetValue(append(msg.Value(), truncatedRecordTerminator...))
+	p.wg.Add(1)
+	select {
+	case p.messages <- msg:
+		return nil
+	default:
+		return ErrDroppedMessage
+	}
 }
 
-// Firehose initializes a producer for Kinesis Firehose with the params supplied in the configuration file
-func (p *Producer) Firehose() (*Producer, error) {
+// Init initalizes a firehose producer with the config file defaults
+func (p *Firehose) Init() (Producer, error) {
 	if conf.Concurrency.Producer < 1 {
 		return nil, ErrBadConcurrency
 	}
 	p.setConcurrency(conf.Concurrency.Producer)
 	p.initChannels()
-	auth, err := authenticate(conf.AWS.AccessKey, conf.AWS.SecretKey)
-	p.kinesis = &kinesis{
-		stream: conf.Firehose.Stream,
-		client: gokinesis.NewWithEndpoint(auth, conf.AWS.Region, fmt.Sprintf(firehoseURL, conf.AWS.Region)),
-	}
+	sess, err := authenticate(conf.AWS.AccessKey, conf.AWS.SecretKey)
+	p.stream = conf.Firehose.Stream
+	p.client = awsFirehose.New(sess)
+	//gokinesis.NewWithEndpoint(auth, conf.AWS.Region, fmt.Sprintf(firehoseURL, conf.AWS.Region)),
```

Comment:
can you please remove the comment code or provide another comment as to why it's commented?